src/net_processing.cpp

```diff
@@ -2151,9 +2193,35 @@ void PeerManagerImpl::RelayTransaction(const uint256& txid, const uint256& wtxid
         // in the announcement.
         if (tx_relay->m_next_inv_send_time == 0s) continue;
 
-        const uint256& hash{peer.m_wtxid_relay ? wtxid : txid};
-        if (!tx_relay->m_tx_inventory_known_filter.contains(hash)) {
-            tx_relay->m_tx_inventory_to_send.insert(hash);
+        bool fanout = true;
+        std::vector<uint256> invs_to_send;
+        if (m_txreconciliation && m_txreconciliation->IsPeerRegistered(peer_id)) {
+            // If this transaction has parents in the mempool and the peer is within the peers with less ancestors
+            // to reconcile, fanout the transaction an all its ancestors. We just add the parents here and leave fanout as true
+            auto it = std::find(fanout_with_ancestors.begin(), fanout_with_ancestors.end(), peer_id);
+            if (it != fanout_with_ancestors.end()) {
+                for (const auto parent_wtxid: parents) {
+                    if (m_txreconciliation->TryRemovingFromSet(peer_id, parent_wtxid)) {
+                        invs_to_send.push_back(parent_wtxid);
+                    }
+                }
+            } else {
+                // If the peer is registered for set reconciliation, maybe pick it as fanout
+                fanout = std::find(fanout_targets.begin(), fanout_targets.end(), peer_id) != fanout_targets.end();
+            }
+        }
+
+        if (fanout || !m_txreconciliation->AddToSet(peer_id, Wtxid::FromUint256(wtxid)).m_succeeded) {
+            // If the transaction cannot be added to the set, we simply fanout, and do not do any special handling here regarding ancestors.
+            // This should not happen under normal conditions, given the set size should be well over the number of transactions received
+            // between reconciling intervals. A peer hitting the limit is likely to be either a broken implementation or an attacker.
+            invs_to_send.push_back(peer->m_wtxid_relay ? wtxid : txid);
+        }
+
+        for (const auto& hash : invs_to_send) {
```

Comment:
If the peer is `m_wtxid_relay=false` should `m_tx_inven tory_known_filter` be checked by `txid` ? Note, given the parent identifiers are retrieved from `GetMemPoolParents()` we can have the identifiers easily. This could avoid some duplicated INV flooding.